### ISSUE TITLE
[Bug] We should recreate Pod/Service due to a map of TFClusterSpec instead of iterations. 

### ISSUE THREAD

**Issue by ScorpioCPH** [caicloud/kubeflow-controller#80]
As described here: https://github.com/caicloud/kubeflow-controller/issues/36
In the function `manageTFJob` (pkg/controller/controller.go) , we recreate Pod/Service in a for loop with `event.Number`:

```go
for i := 0; i < event.Number; i++ {
	if err := c.helper.CreatePod(tfJob, distributedJob.GetSpec(api.TFReplicaWorker, i).Template); err != nil {
		defer runtime.HandleError(err)
		glog.V(2).Infof("Failed creation, decrementing expectations for TFJob %q/%q", tfJob.Namespace, tfJob.Name)
		c.expectations.CreationObserved(jobKey)
		return 0, 0, err
	}
}
```

This maybe recreate the same Pod/Service.
How about use a map of `TFClusterSpec` to cache which we should recreate in the next iteration.

@gaocegege @DjangoPeng WHYT?

**Comment by gaocegege** [comment]
I think it is dup with #51 and we decide to add index to event to solve the problem. I am not sure where to add `TFClusterSpec`, could you give me more details?

**Comment by ScorpioCPH** [comment]
Add `TFClusterStatus` in `TFJob.Status`:

```go
// TFClusterStatus represents a TensorFlow cluster status.
// See: https://www.tensorflow.org/deploy/distributed
//      https://www.tensorflow.org/api_docs/python/tf/train/ClusterSpec
// It is a map from "job_name + task_index" to status (submitted, created, failed):
// {
//     "worker_0": "created",
//     "worker_1": "created",
//     "worker_2": "submitted",
//     "worker_4": "failed",
//     "ps_0": "created",
//     "ps_1": "submitted",
// }
type TFClusterStatus map[string]string
```

**Comment by DjangoPeng** [comment]
LGTM.

**Comment by gaocegege** [comment]
LGTM and maybe we could eliminate the field `TFReplicasStates` since they are dup if you want to add `TFClusterStatus`

**Comment by ScorpioCPH** [comment]
@gaocegege Yes, i'm confusing about `TFReplicaStatus`:
```go
type TFReplicaStatus struct {
	Type *TFReplicaType `json:"type"`
	// State is the overall state of the TFReplica.
	State TFReplicaState `json:"state"`
	// PodNames is the name of all pods.
	PodNames []string `json:"podNames"`
	// ReplicasStates provides the number of TFReplicas in each status.
	TFReplicasStates map[TFReplicaState]int `json:"tfReplicasStates,omitempty"`
}
```
- We keep so many details about a TFJob:
  - why should we keep PodNames in cache, pods come and go but user don't care about this details.
  - TFReplicasStates just tell us how many Replicas of each state, it is hard to maintain and seems useless to user. 
- I think user care about the whole status of a TFJob:
  - Does this Job start training?
  - And the status of each task

**Comment by gaocegege** [comment]
As for the podnames, it is intended to support log-admin and it could be removed now since it is not necessary to know the names of pod.

And TFReplicasStates is migrated from tensorflow/k8s, and I do think it misses some useful info.

I agree with you to add `TFClusterStatus` and the info is useful for users.

**Comment by DjangoPeng** [comment]
> And TFReplicasStates is migrated from tensorflow/k8s, and I do think it misses some useful info.

Agree, and we need to open a PR to tell the upstream the new design of CRD.

WDYT @ScorpioCPH @gaocegege

**Comment by gaocegege** [comment]
@DjangoPeng I think we should file an issue to upstream about our changes in CRD. PR is hard to read, IMO

:thinking:

**Comment by DjangoPeng** [comment]
Yep. Issue is better, and sorry for typo. I think I lose my mind today.

**Comment by gaocegege** [comment]
haha, NVM, SGTM
